File: plugins/google-ads-management/skills/google-ads-bidding-strategy/tests/js_kernel_parity.py
#!/usr/bin/env python3
"""Node-vs-Python parity for THIS skill's js_kernel (bidding_spec.JS_KERNEL).

Computes bidding_core.classify_row expected outputs on the fixture's rows
across several param scenarios (default + two tuned variants exercising
different mismatch branches), then shells out to `node js_kernel_parity.mjs`
to replay the same rows/params through the JS kernel and assert equality.

    python3 tests/js_kernel_parity.py

Dev-only — requires `node` on PATH (no npm packages needed; unlike the shared
tuner harness this script needs no jsdom). Exit 0 = pass, 1 = a failure or
node missing (reported, not silently skipped).
"""
from __future__ import annotations

import json
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE.parent / "scripts"))
sys.path.insert(0, str(HERE.parents[2] / "_shared"))

import bidding_core as core   # noqa: E402
import bidding_spec as spec_mod  # noqa: E402

FIXTURE = HERE / "sample-findings.json"

_FIELDS = ("maturity_score", "recommended_tier", "tier_gap", "mismatch", "severity")


def _expected(universe_rows, params):
    classified = core.classify(universe_rows, params)
    return [{k: r[k] for k in _FIELDS} for r in classified]


def main() -> int:
    node = shutil.which("node")
    if not node:
        sys.stderr.write("ERROR: node not found on PATH — install Node to run this parity check\n")
        return 1

    findings = core.load_findings(str(FIXTURE))
    model = core.compute_model(findings)
    universe = model["_universe"]
    default_params = model["params"]

    tuned_gate = dict(default_params); tuned_gate["conv_gate"] = 15.0
    tuned_bands = dict(default_params)
    tuned_bands["band_edge_1"], tuned_bands["band_edge_2"] = 20.0, 40.0
    tuned_bands["tier_gap_threshold"] = 0.5

    scenarios = [
        {"name": "default params", "params": default_params,
         "expected": _expected(universe, default_params)},
        {"name": "tuned conv_gate (30 -> 15)", "params": tuned_gate,
         "expected": _expected(universe, tuned_gate)},
        {"name": "tuned bands + tighter tier-gap threshold", "params": tuned_bands,
         "expected": _expected(universe, tuned_bands)},
    ]

    rows = spec_mod.html_embed(model)["rows"]  # exactly what the real explorer embeds

    fixture = {"kernel_src": spec_mod.JS_KERNEL, "rows": rows, "scenarios": scenarios}
    with tempfile.TemporaryDirectory() as td:
        fx_path = Path(td) / "fixture.json"
        fx_path.write_text(json.dumps(fixture))
        r = subprocess.run([node, str(HERE / "js_kernel_parity.mjs"), str(fx_path)],
                           capture_output=True, text=True)
        print(r.stdout)
        if r.returncode != 0:
            sys.stderr.write(r.stderr)
            return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: plugins/google-ads-management/skills/google-ads-bidding-strategy/tests/js_kernel_parity.mjs
// Node-vs-Python kernel parity harness for THIS skill's own js_kernel
// (bidding_spec.JS_KERNEL). Deliberately separate from the shared
// skills/google-ads/tests/run_parity.py + tuner_parity.mjs harness (which is
// a frozen file this skill's build is not allowed to edit during the
// parallel M2 batch — see docs/orchestration.md) — the analytics.JS_MIRROR
// portion of this kernel IS already covered by that shared harness via
// analytics_vectors_bidding.json; this script additionally proves the
// skill-specific classify()/summarize() wrapper around it agrees with
// bidding_core.classify_row/summarize on real rows + several param sets.
//
// Usage: node js_kernel_parity.mjs <fixture.json>
// fixture.json: {"kernel_src": "<the JS_KERNEL text>",
//                "rows": [...], "scenarios": [{"name","params","expected":[...]}]}
// Exit 0 = all pass, 1 = a failure.
import fs from "node:fs";

const [, , fixturePath] = process.argv;
if (!fixturePath) {
  console.error("usage: node js_kernel_parity.mjs <fixture.json>");
  process.exit(2);
}
const fx = JSON.parse(fs.readFileSync(fixturePath, "utf8"));

function run(rows, params) {
  let classify, summarize;
  // Direct eval (not indirect) — resolves the pre-declared `let` bindings
  // above via the lexical scope chain, exactly like the real explorer
  // template's inline <script> does with /*__KERNEL__*/.
  // eslint-disable-next-line no-eval
  eval(fx.kernel_src);
  return rows.map((r) => classify(r, params));
}

let failures = 0;
function check(name, cond, detail) {
  console.log(`  ${cond ? "PASS" : "FAIL"}  ${name}` + (detail && !cond ? `  — ${detail}` : ""));
  if (!cond) failures++;
}

function approxEqual(a, b) {
  if (typeof a === "number" && typeof b === "number") return Math.abs(a - b) < 1e-6;
  return JSON.stringify(a) === JSON.stringify(b);
}

for (const scenario of fx.scenarios) {
  console.log(`scenario: ${scenario.name}`);
  const got = run(fx.rows, scenario.params);
  scenario.expected.forEach((exp, i) => {
    const g = got[i];
    for (const key of Object.keys(exp)) {
      check(`row ${i} (${fx.rows[i].campaign}) ${key}`, approxEqual(g[key], exp[key]),
            `got=${JSON.stringify(g[key])} want=${JSON.stringify(exp[key])}`);
    }
    check(`row ${i} (${fx.rows[i].campaign}) block == mismatch`, g.block === g.mismatch);
  });
}

console.log();
if (failures) {
  console.log(`FAILED (${failures})`);
  process.exit(1);
}
console.log("ALL TESTS PASSED");
